File: js/weightedStrings.js
/**
 * Alfabet dari a sampai z memiliki bobot sebesar angka urutannya, 
 * misal: a = 1, b = 2, c = 3, ..., z = 26. 
 * Bobot sebuah string merupakan penjumlahan bobot dari 
 * kesuluruhan karakter. 
 * Contoh: gits = 7 + 9 + 20 + 19 = 55
 */

/**
 * Aturan:
 * Jika terdapat karakter yang berulang dan berurutan perlu dirincikan ke dalam bentuk substring dari perulangan pertama hingga n. 
 * Contoh: bbccc => b, bb, c, cc, ccc. 
 * 
 * Soal: 
 * Buat fungsi untuk menyelesaikan permasalahan Weighted Strings!
 */


function weightedString(s = 'abbcccd', queries = [1, 3, 9, 8]) {
    let weights = new Set();  // Set to store all possible weights
    
    let n = s.length;
    let i = 0;

    // Calculate weights for each character and consecutive substrings
    while (i < n) {
        let currentChar = s[i];
        let weight = currentChar.charCodeAt(0) - 'a'.charCodeAt(0) + 1;  // Get weight of the character
        let sum = 0;
        let count = 0;

        // Loop to handle consecutive occurrences of the same character
        while (i < n && s[i] === currentChar) {
            count++;
            sum += weight;
            weights.add(sum);  // Add each cumulative substring weight to the set
            i++;
        }
    }

    // Process queries
    return queries.map(query => weights.has(query) ? "Yes" : "No");
}

module.exports = weightedString;

File: java/WeightedStrings.java
import java.util.*;

public class WeightedStrings {
    // Fungsi untuk menghitung bobot dan memeriksa query
    public static List<String> weightedStrings(String s, int[] queries) {
        Set<Integer> weights = new HashSet<>();
        int length = s.length();

        // Menghitung bobot semua substring dengan karakter berulang yang berurutan
        for (int i = 0; i < length; ) {
            char currentChar = s.charAt(i);
            int weight = currentChar - 'a' + 1;
            int sum = 0;
            int count = 0;

            while (i < length && s.charAt(i) == currentChar) {
                sum += weight;
                count++;
                weights.add(sum);
                i++;
            }
        }

        // Memeriksa setiap query terhadap bobot yang dihitung
        List<String> results = new ArrayList<>();
        for (int query : queries) {
            if (weights.contains(query)) {
                results.add("Yes");
            } else {
                results.add("No");
            }
        }

        return results;
    }

    // Fungsi utama untuk menguji func weightedStrings
    public static void main(String[] args) {
        String data = "abbbcccd";
        int[] queries = {1, 3, 9, 8};

        List<String> results = weightedStrings(data, queries);
        System.out.println(results);
    }
}
